File: python-ref/gvibu_ref/commands/nl.py
"""nl: number lines of files."""


def run(args: list[str]) -> int:
    files: list[str] = []
    body_start = 1

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == "-v":
            i += 1
            if i >= len(args):
                print("nl: option requires an argument: -v", file=__import__("sys").stderr)
                return 1
            try:
                body_start = int(args[i])
            except ValueError:
                print(f"nl: invalid starting line number: {args[i]}", file=__import__("sys").stderr)
                return 1
        elif arg.startswith("-") and len(arg) > 1:
            print(f"nl: invalid option: {arg}", file=__import__("sys").stderr)
            return 1
        else:
            files.append(arg)
        i += 1

    line_num = body_start
    exit_code = 0

    if not files:
        try:
            exit_code = _number_lines(__import__("sys").stdin, line_num)
        except OSError as e:
            print(f"nl: stdin: {e}", file=__import__("sys").stderr)
            exit_code = 1
    else:
        for fname in files:
            if fname == "-":
                try:
                    exit_code |= _number_lines(__import__("sys").stdin, line_num)
                except OSError as e:
                    print(f"nl: stdin: {e}", file=__import__("sys").stderr)
                    exit_code = 1
            else:
                try:
                    with open(fname) as f:
                        exit_code |= _number_lines(f, line_num)
                except OSError as e:
                    print(f"nl: {fname}: {e}", file=__import__("sys").stderr)
                    exit_code = 1

    return exit_code


def _number_lines(reader, line_num: int) -> int:
    for line in reader:
        line = line.rstrip("\n")
        if not line:
            print()
        else:
            print(f"{line_num:>6}\t{line}")
            line_num += 1
    return 0

File: rust/src/commands/nl.rs
/// nl: number lines of files.
use std::fs::File;
use std::io::{self, BufRead, BufReader, Write};

pub fn run(w: &mut dyn Write, args: &[String]) -> i32 {
    let mut files: Vec<String> = Vec::new();
    let mut body_start = 1usize;

    let mut i = 0;
    while i < args.len() {
        let arg = &args[i];
        if arg == "--" { i += 1; break; }
        if arg == "-v" {
            i += 1;
            if i >= args.len() {
                eprintln!("nl: option requires an argument: -v");
                return 1;
            }
            match args[i].parse::<usize>() {
                Ok(n) => body_start = n,
                Err(_) => {
                    eprintln!("nl: invalid starting line number: {}", args[i]);
                    return 1;
                }
            }
        } else if arg.starts_with('-') && arg.len() > 1 {
            eprintln!("nl: invalid option: {}", arg);
            return 1;
        } else {
            files.push(arg.clone());
        }
        i += 1;
    }

    let mut line_num = body_start;
    let mut exit_code = 0;

    if files.is_empty() {
        let stdin = io::stdin();
        exit_code = number_lines(w, stdin.lock(), &mut line_num);
    } else {
        for fname in files {
            if fname == "-" {
                let stdin = io::stdin();
                exit_code |= number_lines(w, stdin.lock(), &mut line_num);
            } else {
                match File::open(&fname) {
                    Ok(file) => {
                        exit_code |= number_lines(w, BufReader::new(file), &mut line_num);
                    }
                    Err(e) => {
                        eprintln!("nl: {}: {}", fname, e);
                        exit_code = 1;
                    }
                }
            }
        }
    }

    exit_code
}

fn number_lines<R: BufRead>(w: &mut dyn Write, reader: R, line_num: &mut usize) -> i32 {
    for line in reader.lines() {
        match line {
            Ok(l) => {
                if l.is_empty() {
                    // Empty lines: don't number, just output
                    pwriteln!(w);
                } else {
                    pwriteln!(w, "{:>6}\t{}", line_num, l);
                    *line_num += 1;
                }
            }
            Err(e) => {
                eprintln!("nl: read error: {}", e);
                return 1;
            }
        }
    }
    0
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::sink;

    #[test]
    fn test_nl_dev_null() {
        assert_eq!(run(&mut sink(), &["/dev/null".into()]), 0);
    }

    #[test]
    fn test_nl_no_args() {
        assert_eq!(run(&mut sink(), &[]), 0);
    }

    #[test]
    fn test_nl_invalid_option() {
        assert_eq!(run(&mut sink(), &["-x".into()]), 1);
    }
}
